File: 06 Shell/testUtils.js
function randomArray(n) {
    const arr = [];
    for (let i = 0; i < n; i++) {
        arr.push(Math.random());
    }
    return arr;
}

function sortedArray(n) {
    const arr = [];
    for (let i = 0; i < n; i++) {
        arr.push(i);
    }
    return arr;
}

function reversedArray(n) {
    const arr = [];
    for (let i = 0; i < n; i++) {
        arr.push(n - i - 1);
    }
    return arr;
}

// n - сколько элементов надо перемешать в массиве
function shuffle(arr, n) {
    const indexes = new Set();

    while (indexes.size < n) {
        indexes.add(Math.floor(Math.random() * arr.length));
    }

    const values = indexes.values();
    let index = values.next().value;
    const value = arr[index];
    for (let i of values) {
        arr[index] = arr[i];
        index = i;
    }
    arr[index] = value;
}

module.exports = {
    randomArray,
    sortedArray,
    reversedArray,
    shuffle
};

File: go/testutils/testutils.go
package testutils

import (
	"math/rand"
	"time"
)

func Bench(fn func()) time.Duration {
	start := time.Now()
	fn()
	return time.Since(start)
}

func RandomArray(n int) []float64 {
	randomSource := rand.NewSource(time.Now().UnixNano())
	random := rand.New(randomSource)

	data := make([]float64, n)
	for i := 0; i < len(data); i++ {
		data[i] = random.Float64()
	}

	return data
}

func ConstantArray(n int, value float64) []float64 {
	data := make([]float64, n)
	for i := 0; i < len(data); i++ {
		data[i] = value
	}

	return data
}

func Shuffle(arr []float64, n int) {
	randomSource := rand.NewSource(time.Now().UnixNano())
	random := rand.New(randomSource)

	indexes := make(map[int]bool)
	for len(indexes) < n {
		indexes[random.Intn(len(arr))] = true
	}

	var index int
	var value float64

	for k := range indexes {
		index = k
		value = arr[index]
		break
	}
	delete(indexes, index)

	for k := range indexes {
		arr[index] = arr[k]
		index = k
	}
	arr[index] = value
}

func CopyArray(arr []float64) []float64 {
	result := make([]float64, len(arr))
	copy(result, arr)
	return result
}
